### PR TITLE
[Neo Core Fix]use strong randomness

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This pr adopts strong randomness to generate random numbers.  I am not sure if we can consider this as a bug, so i will mark it as an optimization.

Fixes # https://github.com/neo-project/neo/issues/3431

## Type of change

- [x] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by cschuchardt88 — COMMENTED** [review]
Aww you beat me to it. At least use this one, its stronger and more random.
```csharp
static ulong GetNonce()
{
    Span<byte> seed = stackalloc byte[8];
    Span<byte> source = stackalloc byte[8];
    RandomNumberGenerator.Fill(seed);
    RandomNumberGenerator.Fill(source);
    var buffer = HMACSHA384.HashData(seed, source);
    var start = Math.Abs(new Random(Random.Shared.Next()).Next(buffer.Length - 8));
    return BinaryPrimitives.ReadUInt64LittleEndian(buffer[start..]);
}
```

**Comment by dusmart** [comment]
> its stronger and more random.

Sorry. Why? is there any proof?

**Comment by cschuchardt88** [comment]
> > its stronger and more random.
> 
> Sorry. Why? is there any proof?

Random number generator can be predicted, so hashing it will be more secure. If that isn't obvious than I'm sorry.

Just because it has a name of `RandomNumberGenerator` in the `System.Security.Cryptography` namespace of `dotnet` doesn't mean it can't be done.

**Comment by dusmart** [comment]
Hi [cschuchardt88](https://github.com/cschuchardt88),

Thank you for your input. I appreciate your detailed explanation regarding the use of hashing to enhance security. You make valid points about the predictability of random number generators and the benefits of HMACSHA384 in ensuring data integrity and resistance to attacks.

However, I have some questions and concerns about the implementation:

Randomness and Security: While I understand that random number generators can be predictable, the System.Security.Cryptography.RandomNumberGenerator class is designed to provide cryptographically secure random numbers. If there are specific known vulnerabilities with this class, could you please point me to the relevant resources or documentation?

Performance: Introducing hashing, especially with HMACSHA384, adds computational overhead. Have you considered the performance impact of this change, particularly in high-throughput applications?

Random Part of the Hash: The approach of selecting a random part of the hash to use as a nonce introduces additional randomness. However, the Random.Shared.Next() method itself relies on a pseudo-random number generator. Could this approach inadvertently reduce the overall security?

Simplicity and Complexity: The original method is straightforward and easy to understand. While adding complexity can enhance security, it's essential to balance it with maintainability and clarity. Could we achieve the desired security improvements with a simpler approach?

Again, I appreciate your efforts to improve the security of our code. If you have further insights or resources, I would be eager to learn more.

Best regards

**Comment by cschuchardt88** [comment]
[eoas2004.pdf](https://github.com/user-attachments/files/16319467/eoas2004.pdf)

**Comment by cschuchardt88** [comment]
>vulnerabilities in `System.Security.Cryptography.RandomNumberGenerator`

I haven't heard of any.

>Performance: Introducing hashing, especially with HMACSHA384, adds computational overhead. Have you considered the performance impact of this change?

The impact with hashing algorithms are very small unlike encrypting data. Since we dealing with small amount of data here the impact will so small you couldn't measure it probably.

>Random Part of the Hash: The approach of selecting a random part of the hash to use as a nonce introduces additional randomness. However, the Random.Shared.Next() method itself relies on a pseudo-random number generator. Could this approach inadvertently reduce the overall security?

No, we are using a small amount of the hash already, and if the number is the same you wouldn't know which hash it came from, because different hashes may have the same part so a collision here is a good thing.

>Again, I appreciate your efforts to improve the security of our code. If you have further insights or resources, I would be eager to learn more.

https://stackoverflow.com/questions/1257299/why-use-the-c-sharp-class-system-random-at-all-instead-of-system-security-crypto

Random Number Generation paper here 
[eoas2004.pdf](https://github.com/user-attachments/files/16319513/eoas2004.pdf)


>Additionally, I noticed that your response had a bit of a sharp tone. For our team and community to function well, it's crucial that we communicate respectfully and constructively. I value your expertise and contributions, and I'm sure our discussions will be more productive if we maintain a courteous and collaborative atmosphere.

Yes may of been proceed that way, however I don't mean any disrespectful by it.

**Comment by Jim8y** [comment]
`RandomNumberGenerator` already provides a secure enough solution for our usage. We are not doing a random number generator research here. So, not necessary to go that far.  And BTW, we will replace all of these with BLS in the future, so, chill.

File: src/Plugins/DBFTPlugin/Consensus/ConsensusContext.MakePayload.cs [hunk 1]
```diff
@@ -166,12 +166,22 @@ public ExtensiblePayload MakePrepareResponse()
...
-            Random _random = new();
             Span<byte> buffer = stackalloc byte[8];
-            _random.NextBytes(buffer);
+            using (var rng = System.Security.Cryptography.RandomNumberGenerator.Create())
+            {
+                rng.GetBytes(buffer);
+            }
             return BinaryPrimitives.ReadUInt64LittleEndian(buffer);
         }
+
```
Comment (shargon) [inline src/Plugins/DBFTPlugin/Consensus/ConsensusContext.MakePayload.cs:185]:
```suggestion
```

**Review by shargon — APPROVED** [review]

**Review by cschuchardt88 — APPROVED** [review]

**Pull request merged**